staging/src/k8s.io/apiserver/pkg/cel/ip.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"math"
+	"net/netip"
+	"reflect"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// IP provides a CEL representation of an IP address.
+type IP struct {
+	netip.Addr
+
+	// Raw is the original string representation of the IP address.
+	// This is used so that we can compare the original string representation
+	// with the parsed IP address.
+	Raw string
+}
+
+var (
+	IPObject        = decls.NewObjectType("kubernetes.IP")
+	ipAddrTypeValue = types.NewTypeValue("kubernetes.IP")
+	IPType          = cel.ObjectType("kubernetes.IP")
```

Comment:
If the `ip` type doesn't have any accessible fields, prefer to treat it as an opaque type whose properties can only be accessed via accessors, e.g. `cel.OpaqueType("kubernetes.IP")`